### ISSUE TITLE
Wazuh agent 3.7.2 crashing

### ISSUE THREAD

**Issue by CharlieStokes** [wazuh/wazuh#4360]
I've installed version 3.7.2 of the Wazuh agent software on 5 aws Windows instances (Server 2016) I've registered them with the agent manually and added in the authentication keys. 

The Wazuh service starts up fine, and shows as connected on the manager. However within a few seconds the service stops, there are no errors in the ossec.log file. However the following is written out to the event log. 

Faulting application name: ossec-agent.exe, version: 0.0.0.0, time stamp: 0x5c2f2e53
Faulting module name: msvcrt.dll, version: 7.0.14393.2457, time stamp: 0x5b7e2dd0
Exception code: 0xc0000005
Fault offset: 0x0008918c
Faulting process id: 0x3060
Faulting application start time: 0x01d5b01bd4ebc199
Faulting application path: C:\Program Files (x86)\ossec-agent\ossec-agent.exe
Faulting module path: C:\Windows\System32\msvcrt.dll
Report Id: 35e617a5-c51c-4b5f-8093-4d9131fe8971
Faulting package full name: 
Faulting package-relative application ID: 

I've only made one change to ossec.conf switching the port o tcp. Any information on how I can diagnose this issue?

**Comment by vikman90** [comment]
Hi @CharlieStokes,

This issue is close to #2467 (solved in #2486, v3.9.0), but I cannot confirm that it's the same bug.

This is a key clue:

> Faulting module name: msvcrt.dll, version: 7.0.14393.2457, time stamp: 0x5b7e2dd0
> Fault offset: 0x0008918c

If we take a look at the Visual C Library (_msvcrt.dll_ file) with version 7.0.14393.2457, this is what we find at address 0x00089180:

<img width="1242" alt="image" src="https://user-images.githubusercontent.com/10536251/70676441-0628a380-1c41-11ea-942c-8bea7cc7ce36.png">

Address 0x0008918c is between 0x00089180 and 0x00089210, so I'm sure that the agent crashed while executing the `strlen()` function, and that's why I think this issue may be related to #2467.

On the other hand, having the agent crashing only in TCP mode makes me remember issue #951 (solved in #2329, v3.8.0).

## How to diagnose

1. Could you confirm that this problem appears only in TCP? I mean, does the agent consistently crash in TCP while it works in UDP?
2. The last logs before the agent dies could help out: Please enable debugging [`windows.debug=2`](https://documentation.wazuh.com/3.10/user-manual/reference/internal-options.html#windows) and let us know what the agent prints before crashing.
3. Version 3.7.2 is one year old, I would suggest you upgrade your agent to 3.10.2 or 3.9.5 as this bug might be already fixed.

Hope this helps.

Best regards.

**Comment by CharlieStokes** [comment]
Thanks for the reply, i've upgraded one of the Windows instances to version 3.9.5 and 3.10.2 both of which still exhibit the same issue. Our Wazuh Manager is currently configured to use TCP, we also have 300 + Centos and RHEL instances connecting to using TCP with no issues. 

As a result testing UDP might be tricky. I've added the following to the ossec.conf - hopefully that's correct?

  <Windows>
	<windows.debug>2</windows.debug>
  </Windows>

The agent logs are attached: 
[wazuh-logs.txt](https://github.com/wazuh/wazuh/files/3954918/wazuh-logs.txt)

Same error appears in the event log: 

Faulting application name: ossec-agent.exe, version: 0.0.0.0, time stamp: 0x5d88e0a5
Faulting module name: msvcrt.dll, version: 7.0.14393.2457, time stamp: 0x5b7e2dd0
Exception code: 0xc0000005
Fault offset: 0x0008918c
Faulting process id: 0x3638
Faulting application start time: 0x01d5b0cde9ef889a
Faulting application path: C:\Program Files (x86)\ossec-agent\ossec-agent.exe
Faulting module path: C:\Windows\System32\msvcrt.dll
Report Id: e116e617-c07d-4cee-b44b-d0e9cdb0ff25
Faulting package full name: 
Faulting package-relative application ID:

**Comment by vikman90** [comment]
@CharlieStokes Sorry to read this.

I suggested you enable _debug=2_ as you can see in the link of the option name. You can enable debug adding this line to the file _C:\Program Files (x86)\ossec-agent\local_internal_options.conf_:

```
windows.debug=2
```

*Warning*: this configuration makes the agent write a huge amount of data into _ossec.log_. Remember to disable debug back when you pick up the logs.

We will do our best to reproduce this crash and fix that ASAP. However, we have not seen a crash recently. I wonder if you might have a special configuration that discovers this bug.

We would greatly appreciate if you could share your agent configuration (agent's _ossec.conf_) and the shared configuration (_/var/ossec/etc/shared/*/agent.conf_ on the manager).

I think of these steps to address the problem:

1. Does your `<address>` stanza contain an IP or a hostname?
2. Are you running 3.9.5 / 3.10.2 on the manager?
3. Per the time from getting connected to the crash, I wonder if the problem might be related with the shared configuration transfer. Could you confirm that the content of the agent's folder `shared` matches the shared configuration of the manager?
4. This log makes me think that there is a mistake in the Rootcheck's configuration as it's referencing files that should only be on the UNIX version of the agent:
```
ERROR: No rootcheck_files file: '/var/ossec/etc/rootcheck/rootkit_files.txt'
```
5. Could you selectively disable components in order to find out which of them is producing the problem?
- Disable SCA:
```xml
<sca>
  <enabled>no</enabled>
</sca>
```
- Disable Syscheck:
```xml
<sycheck>
  <disabled>yes</disabled>
</syscheck>
```
- Disable Rootcheck:
```xml
<rootcheck>
  <disabled>yes</disabled>
</rootcheck>
```
- Disable log monitoring: Remove or comment `<localfile>` stanzas.
- Disable Syscollector:
```xml
<wodle name="syscollector">
  <disabled>yes</disabled>
</wodle>
```

If all this provides us with no data enough to reproduce the issue, I ask you if you would be able to run a debugging version of the agent —we would provide you with a version of the agent that includes debugging data— with the Dr. Memory application. This program helps to find runtime errors including the source code.

Thank you.
Best regards.

**Comment by CharlieStokes** [comment]
Hi again.

I've done some further investigation, i've attached the ossec and agent conf files as well as the logs with debug enabled. 

[agent.conf.txt](https://github.com/wazuh/wazuh/files/3960989/agent.conf.txt)
[ossec.conf.txt](https://github.com/wazuh/wazuh/files/3960990/ossec.conf.txt)
[ossec-debug.txt](https://github.com/wazuh/wazuh/files/3960991/ossec-debug.txt)

You can see from the config file that i'm using the IP address of the manager in the configuration file. We are currently running wazuh-manager-3.10.2-1.

Not quite sure what you mean by checking the 'shared' folder on the manager? I've tried disabling each section as you have mentioned below, however i'm still getting the same issue. The agent crashes a few seconds after starting.

**Comment by vikman90** [comment]
Hi,

So sorry, I've not managed to let the agent crash with your configuration and the same _msvcrt.dll_ version as you. If the agent still crashes even disabling every component, the bug must be in the communication part (agent component itself). That is very helpful.

Exactly, I meant the _agent.conf_ by the shared configuration.

I'm trying to use [Dr. Memory](https://drmemory.org) to debug the agent:

```
"C:\Program Files (x86)\Dr. Memory\bin\drmemory.exe" "C:\Program Files (x86)\ossec-agent\ossec-agent.exe" start
```

### Dr. Memory report

[dr_memory.txt](https://github.com/wazuh/wazuh/files/3970885/dr_memory.txt)

Maybe this report has the key to detect the bug that is producing the crash, but I have to check the items. I'm pretty sure many of them are either false positives or issues in the Microsoft Visual C Runtime library.

I wonder if you could run Dr. Memory in your system to check if it reports the crash, that would be very helpful.

Thank you.
Regards.

**Comment by bah07** [comment]
Hi @vikman90 @CharlieStokes, I think we've found the problem.

I have seen that in the configuration is being applied, through centralized configuration, the configuration of a Linux system.
Among the paths that are monitored are `\boot`. This directory is being monitored because the current working directory in Windows is `C:` therefore the function `opendir()` return successfully because this directory exists and is correct `c:\boot`. 
 The problem comes in the algorithm that detects deleted files in programmed scans when scrolling through the monitored files, in Windows, it is checked that the path contains the character `:` but the stored path does not contain this character.

https://github.com/wazuh/wazuh/blob/f38516ad45643287e4ee4e781b57cee6129c4df3/src/syscheckd/create_db.c#L137-L140

The problem is solved by simply adding tag `os="Linux"` in the centralized configuration in the agent.conf so that it only applies to the Linux agents.
```
<agent_config os="Linux">

  <!-- Shared agent configuration here -->
  <syscheck>
    <!-- Directories to check  (perform all possible verifications) -->
    <directories check_all="yes">/etc,/usr/bin,/usr/sbin</directories>
    <directories check_all="yes">/bin,/sbin,/boot</directories>
  </syscheck>
</agent_config>
```
We are currently making many changes in the FIM module and this error is controlled because when a relative path is introduced we get the full path.

@CharlieStokes Could you please check by omitting the /boot directory from the configuration in the way I'm telling you above solves the problem? 

Thank you for reporting this bug and any problem you have do not hesitate to contact us.

**Comment by CharlieStokes** [comment]
Hi
Thanks for all the information on this one - I've applied the config change suggested above. 

The good news is the agent is now stable and no longer crashes on the initial scan. I've commented out both of the <syscheck> sections that contain references to /etc /usr/bin etc. 

However I may need to raise another issue as I'm now seeing a new error message in the logs which may be unrelated to this change. Previously these Windows agents were able to connect to the manager instance with no issues, however now the logs contain: 

ERROR: (1218): Unable to send message to 'server': An established connection was aborted by the software in your host machine.

**Comment by bah07** [comment]
Hi @CharlieStokes,
Sorry, I took so long to respond, I didn't notice your comment.

It may be already late, but most likely by re-registering the agent the problem will be solved. 

I will proceed to close this issue because the problem referred to first was solved in release 3.12.

If the error you made in the previous comment persists, please open a new issue with the necessary information and we will be happy to help.

Again, apologies for the late response. Best regards.
